euicc/logger.c

```diff
@@ -0,0 +1,74 @@
+#include "logger.h"
+
+#include <unistd.h>
+
+// Refs: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html
+#define COLOR_RED 31
+#define COLOR_GREEN 32
+#define COLOR_CLEAR 0
+
+static void colorize(FILE *fp, const int color) {
+#if _WIN32
+    // https://devblogs.microsoft.com/commandline/windows-command-line-introducing-the-windows-pseudo-console-conpty/
+    // Minimum supported version: Windows 10 version 1809
+    if (_isatty(_fileno(fp)))
+#else
+    if (isatty(fileno(fp)))
+#endif
+        fprintf(fp, "\033[0;%dm", color);
+}
+
+static void euicc_hex_print(FILE *fp, const uint8_t *data, const uint32_t length) {
+    for (uint32_t i = 0; i < length; i++)
+        fprintf(fp, " %02X", data[i] & 0xFF);
+}
+
+inline void euicc_apdu_request_print(FILE *fp, const struct apdu_request *req, const uint32_t req_len) {
+    if (fp == NULL)
+        return;
+    colorize(fp, COLOR_GREEN);
+    fprintf(fp, "[DEBUG] [APDU] [TX] CLA: %02X, INS: %02X, P1: %02X, P2: %02X, Lc: %02X, Data:", req->cla, req->ins,
+            req->p1, req->p2, req->length);
+    euicc_hex_print(fp, req->data, req_len - sizeof(struct apdu_request));
+    colorize(fp, COLOR_CLEAR);
+    fputc('\n', fp);
```

Comment:
If you can do output in one time, don't split it into multiple statements. Otherwise, it may be interleaved in a multithreaded environment or in a shared fd environment. Since libeuicc is a library, we SHALL consider this issue.

Reply:
You can construct the whole string using snprintf, and output it using one call.

Reply:
<img width="1148" height="278" alt="CleanShot 2025-09-07 at 23 02 22@2x" src="https://github.com/user-attachments/assets/496a78c6-a152-4b21-a7b4-01f0d734f779" />